File: src/tickoni/c_abi/wksp.zig
/// Narrow Zig bindings over the subset of src/util/wksp/fd_wksp.h needed to
/// create/join a named, process-shared workspace that backs Tango
/// mcache/dcache/fseq/cnc objects for V1.14 process mode.
///
/// Uses FD_SHMEM_NORMAL_PAGE_SZ (4 KiB) workspaces, which fd_wksp_new_named
/// backs with a plain file under the FD_SHMEM_PATH directory (see
/// src/util/shmem/fd_shmem_admin.c) rather than hugetlbfs — no gigantic/huge
/// pages or root privileges are required, matching the roadmap requirement
/// that Tickoni not inherit Firedancer validator's hugepage assumption.
///
/// Link requirements: -lfd_util plus shim/wksp.c at link time.
const std = @import("std");

// ---------------------------------------------------------------------------
// Constants (from fd_shmem.h)
// ---------------------------------------------------------------------------

pub const shmem_normal_page_sz: usize = 4096;

/// Opaque workspace handle returned by fd_wksp_new_named / fd_wksp_attach.
pub const Wksp = opaque {};

// ---------------------------------------------------------------------------
// Extern declarations — require -lfd_util at link time
// ---------------------------------------------------------------------------

pub const wksp_success: c_int = 0;

/// Returns FD_WKSP_SUCCESS (0) or a negative FD_WKSP_ERR_* code — NOT a
/// workspace handle. Call wkspAttach afterward to join it.
extern fn tk_wksp_new_named(
    name: [*:0]const u8,
    page_sz: usize,
    sub_cnt: usize,
    sub_page_cnt: [*]const usize,
    sub_cpu_idx: [*]const usize,
    mode: usize,
    seed: u32,
    opt_part_max: usize,
) c_int;
extern fn tk_wksp_delete_named(name: [*:0]const u8) c_int;
extern fn tk_wksp_attach(name: [*:0]const u8) ?*Wksp;
extern fn tk_wksp_detach(wksp: *Wksp) c_int;
extern fn tk_wksp_alloc_at_least(
    wksp: *Wksp,
    alignment: usize,
    sz: usize,
    tag: usize,
    lo: *usize,
    hi: *usize,
) usize;
extern fn tk_wksp_alloc(wksp: *Wksp, alignment: usize, sz: usize, tag: usize) usize;
extern fn tk_wksp_free(wksp: *Wksp, gaddr: usize) void;
extern fn tk_wksp_laddr(wksp: *const Wksp, gaddr: usize) ?*anyopaque;
extern fn tk_wksp_gaddr(wksp: *const Wksp, laddr: *const anyopaque) usize;

pub fn wkspNewNamed(
    name: [*:0]const u8,
    page_sz: usize,
    sub_cnt: usize,
    sub_page_cnt: [*]const usize,
    sub_cpu_idx: [*]const usize,
    mode: usize,
    seed: u32,
    opt_part_max: usize,
) c_int {
    return tk_wksp_new_named(name, page_sz, sub_cnt, sub_page_cnt, sub_cpu_idx, mode, seed, opt_part_max);
}

pub fn wkspDeleteNamed(name: [*:0]const u8) c_int {
    return tk_wksp_delete_named(name);
}

pub fn wkspAttach(name: [*:0]const u8) ?*Wksp {
    return tk_wksp_attach(name);
}

pub fn wkspDetach(wksp: *Wksp) c_int {
    return tk_wksp_detach(wksp);
}

pub fn wkspAllocAtLeast(wksp: *Wksp, alignment: usize, sz: usize, tag: usize, lo: *usize, hi: *usize) usize {
    return tk_wksp_alloc_at_least(wksp, alignment, sz, tag, lo, hi);
}

/// Mirrors the upstream workspace allocation helper: a fixed-alignment
/// wrapper around fd_wksp_alloc_at_least with dummy [lo,hi) outputs.
pub fn wkspAlloc(wksp: *Wksp, alignment: usize, sz: usize, tag: usize) usize {
    return tk_wksp_alloc(wksp, alignment, sz, tag);
}

pub fn wkspFree(wksp: *Wksp, gaddr: usize) void {
    tk_wksp_free(wksp, gaddr);
}

pub fn wkspLaddr(wksp: *const Wksp, gaddr: usize) ?*anyopaque {
    return tk_wksp_laddr(wksp, gaddr);
}

pub fn wkspGaddr(wksp: *const Wksp, laddr: *const anyopaque) usize {
    return tk_wksp_gaddr(wksp, laddr);
}

// ---------------------------------------------------------------------------
// Tests — constants only; no C linkage required
// ---------------------------------------------------------------------------

test "shmem normal page size matches header" {
    try std.testing.expectEqual(@as(usize, 4096), shmem_normal_page_sz);
}

File: src/tickoni/c_abi/shim/wksp.c
/* Thin wrappers around Firedancer workspace primitives. */

#include "../../../util/fd_util.h"
#include "../../../util/wksp/fd_wksp.h"

int
tk_wksp_new_named( char const *  name,
                   ulong         page_sz,
                   ulong         sub_cnt,
                   ulong const * sub_page_cnt,
                   ulong const * sub_cpu_idx,
                   ulong         mode,
                   uint          seed,
                   ulong         opt_part_max ) {
  return fd_wksp_new_named( name, page_sz, sub_cnt, sub_page_cnt, sub_cpu_idx, mode, seed, opt_part_max );
}

int tk_wksp_delete_named( char const * name ) { return fd_wksp_delete_named( name ); }
fd_wksp_t * tk_wksp_attach( char const * name ) { return fd_wksp_attach( name ); }
int tk_wksp_detach( fd_wksp_t * wksp ) { return fd_wksp_detach( wksp ); }
ulong tk_wksp_alloc_at_least( fd_wksp_t * wksp, ulong alignment, ulong sz, ulong tag, ulong * lo, ulong * hi ) { return fd_wksp_alloc_at_least( wksp, alignment, sz, tag, lo, hi ); }
ulong tk_wksp_alloc( fd_wksp_t * wksp, ulong alignment, ulong sz, ulong tag ) { return fd_wksp_alloc( wksp, alignment, sz, tag ); }
void tk_wksp_free( fd_wksp_t * wksp, ulong gaddr ) { fd_wksp_free( wksp, gaddr ); }
void * tk_wksp_laddr( fd_wksp_t const * wksp, ulong gaddr ) { return fd_wksp_laddr( wksp, gaddr ); }
ulong tk_wksp_gaddr( fd_wksp_t const * wksp, void const * laddr ) { return fd_wksp_gaddr( wksp, laddr ); }
int tk_wksp_exists_named( char const * name ) { return !fd_shmem_info( name, 0UL, NULL ); }
